File: src/main/java/com/checkmarx/falsepositive/xss/reflected/ReflectedXSS_FP_WhitelistValidation.java
package com.checkmarx.falsepositive.xss.reflected;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.List;

/**
 * FALSE POSITIVE SCENARIOS: Reflected XSS - Whitelist Validation Patterns
 * 
 * Pattern: User input validated against a fixed whitelist of allowed values
 * 
 * Whitelist validation ensures only known-safe values are used.
 * CxQL doesn't recognize whitelist checks as sanitization.
 * 
 * All scenarios are FALSE POSITIVES - SAFE code that should NOT be flagged.
 */
public class ReflectedXSS_FP_WhitelistValidation {
    
    private static final Set<String> ALLOWED_CATEGORIES = new HashSet<>(
        Arrays.asList("electronics", "books", "clothing", "food", "toys")
    );
    
    private static final List<String> ALLOWED_COLORS = Arrays.asList(
        "red", "green", "blue", "yellow", "orange", "purple"
    );
    
    private static final String[] ALLOWED_SIZES = {"XS", "S", "M", "L", "XL", "XXL"};
    
    /**
     * FALSE POSITIVE: Set.contains() whitelist check
     * Only values in the HashSet are allowed
     */
    public void showSetWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String category = request.getParameter("category");
        
        PrintWriter out = response.getWriter();
        if (ALLOWED_CATEGORIES.contains(category)) {
            out.write("Category: " + category); // FALSE POSITIVE - Whitelisted value
        } else {
            out.write("Invalid category");
        }
    }
    
    /**
     * FALSE POSITIVE: List.contains() whitelist check
     * Only values in the List are allowed
     */
    public void showListWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String color = request.getParameter("color");
        
        PrintWriter out = response.getWriter();
        if (ALLOWED_COLORS.contains(color)) {
            out.write("Color: " + color); // FALSE POSITIVE - Whitelisted value
        } else {
            out.write("Invalid color");
        }
    }
    
    /**
     * FALSE POSITIVE: Array contains check
     * Only values in the array are allowed
     */
    public void showArrayWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String size = request.getParameter("size");
        
        boolean isValid = Arrays.asList(ALLOWED_SIZES).contains(size);
        
        PrintWriter out = response.getWriter();
        if (isValid) {
            out.write("Size: " + size); // FALSE POSITIVE - Whitelisted value
        } else {
            out.write("Invalid size");
        }
    }
    
    /**
     * FALSE POSITIVE: Switch statement whitelist
     * Only specific case values are processed
     */
    public void showSwitchWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String status = request.getParameter("status");
        
        PrintWriter out = response.getWriter();
        switch (status) {
            case "pending":
            case "approved":
            case "rejected":
            case "completed":
                out.write("Status: " + status); // FALSE POSITIVE - Whitelisted by switch
                break;
            default:
                out.write("Invalid status");
        }
    }
    
    /**
     * FALSE POSITIVE: Enum.valueOf() validation
     * Only valid enum values pass
     */
    public void showEnumWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String priority = request.getParameter("priority");
        
        PrintWriter out = response.getWriter();
        try {
            Priority p = Priority.valueOf(priority.toUpperCase());
            out.write("Priority: " + p.name()); // FALSE POSITIVE - Valid enum value
        } catch (IllegalArgumentException e) {
            out.write("Invalid priority");
        }
    }
    
    /**
     * FALSE POSITIVE: Stream filter whitelist
     * Filtered through stream predicate
     */
    public void showStreamWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String type = request.getParameter("type");
        
        boolean isValid = ALLOWED_CATEGORIES.stream()
            .anyMatch(allowed -> allowed.equals(type));
        
        PrintWriter out = response.getWriter();
        if (isValid) {
            out.write("Type: " + type); // FALSE POSITIVE - Stream filtered
        } else {
            out.write("Invalid type");
        }
    }
    
    /**
     * FALSE POSITIVE: equalsIgnoreCase whitelist
     * Explicit string comparison
     */
    public void showExplicitWhitelist(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String action = request.getParameter("action");
        
        PrintWriter out = response.getWriter();
        if ("view".equalsIgnoreCase(action) || 
            "edit".equalsIgnoreCase(action) || 
            "delete".equalsIgnoreCase(action)) {
            out.write("Action: " + action); // FALSE POSITIVE - Explicit whitelist
        } else {
            out.write("Invalid action");
        }
    }
    
    // Enum for whitelist validation
    private enum Priority {
        LOW, MEDIUM, HIGH, CRITICAL
    }
}

File: src/main/csharp/FalsePositive/XSS/ReflectedXSS_FP_WhitelistValidation.cs
using System;
using System.Web;
using System.Web.UI;
using System.Collections.Generic;
using System.Linq;

namespace Checkmarx.FalsePositive.XSS
{
    /// <summary>
    /// FALSE POSITIVE SCENARIOS: Reflected XSS - Whitelist Validation (C#)
    /// 
    /// Pattern: User input validated against a fixed whitelist of allowed values
    /// 
    /// Whitelist validation ensures only known-safe values are used.
    /// CxQL doesn't recognize these patterns as sanitization.
    /// 
    /// All scenarios are FALSE POSITIVES - SAFE code that should NOT be flagged.
    /// </summary>
    public class ReflectedXSS_FP_WhitelistValidation : Page
    {
        private static readonly HashSet<string> AllowedCategories = new HashSet<string>
        {
            "electronics", "books", "clothing", "food", "toys"
        };
        
        private static readonly List<string> AllowedColors = new List<string>
        {
            "red", "green", "blue", "yellow", "orange", "purple"
        };
        
        private static readonly string[] AllowedSizes = { "XS", "S", "M", "L", "XL", "XXL" };
        
        /// <summary>
        /// FALSE POSITIVE: HashSet.Contains() whitelist check
        /// </summary>
        protected void ShowHashSetWhitelist()
        {
            string category = Request.QueryString["category"];
            
            if (AllowedCategories.Contains(category))
            {
                Response.Write("Category: " + category); // FALSE POSITIVE - Whitelisted
            }
            else
            {
                Response.Write("Invalid category");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: List.Contains() whitelist check
        /// </summary>
        protected void ShowListWhitelist()
        {
            string color = Request.QueryString["color"];
            
            if (AllowedColors.Contains(color))
            {
                Response.Write("Color: " + color); // FALSE POSITIVE - Whitelisted
            }
            else
            {
                Response.Write("Invalid color");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Array.Contains() whitelist check
        /// </summary>
        protected void ShowArrayWhitelist()
        {
            string size = Request.QueryString["size"];
            
            if (AllowedSizes.Contains(size))
            {
                Response.Write("Size: " + size); // FALSE POSITIVE - Whitelisted
            }
            else
            {
                Response.Write("Invalid size");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Switch statement whitelist
        /// </summary>
        protected void ShowSwitchWhitelist()
        {
            string status = Request.QueryString["status"];
            
            switch (status)
            {
                case "pending":
                case "approved":
                case "rejected":
                case "completed":
                    Response.Write("Status: " + status); // FALSE POSITIVE - Switch whitelist
                    break;
                default:
                    Response.Write("Invalid status");
                    break;
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Enum.TryParse() validation
        /// </summary>
        protected void ShowEnumWhitelist()
        {
            string priority = Request.QueryString["priority"];
            
            if (Enum.TryParse<Priority>(priority, true, out Priority p))
            {
                Response.Write("Priority: " + p.ToString()); // FALSE POSITIVE - Valid enum
            }
            else
            {
                Response.Write("Invalid priority");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: LINQ Any() whitelist
        /// </summary>
        protected void ShowLinqWhitelist()
        {
            string type = Request.QueryString["type"];
            
            if (AllowedCategories.Any(a => a.Equals(type, StringComparison.OrdinalIgnoreCase)))
            {
                Response.Write("Type: " + type); // FALSE POSITIVE - LINQ filtered
            }
            else
            {
                Response.Write("Invalid type");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Explicit string comparison
        /// </summary>
        protected void ShowExplicitWhitelist()
        {
            string action = Request.QueryString["action"];
            
            if (action == "view" || action == "edit" || action == "delete")
            {
                Response.Write("Action: " + action); // FALSE POSITIVE - Explicit whitelist
            }
            else
            {
                Response.Write("Invalid action");
            }
        }
        
        private enum Priority
        {
            Low,
            Medium,
            High,
            Critical
        }
    }
}

